keps/sig-scheduling/3521-pod-schedule-readiness/README.md

```diff
@@ -0,0 +1,1053 @@
+Pod for scheduling. The pattern for this story would be to use a mutating webhook to force creating
+pods in a "not-ready to schedule" state that the custom orchestrator changes to ready at a later
+time based on its own evaluations.
+
+#### Story 2
+
+This story is an extension of the previous one in that more than one custom orchestrator could be
+deployed on the same cluster, therefore they want an API that enables establishing an agreement om
+when a pod is considered ready for scheduling.
+
+#### Story 3
+
+As an advanced scheduler developer, I want to compose a series of scheduler PreEnqueue plugins
+to guard the schedulability of my Pods. This enables splitting custom enqueue admission logic into
+several building blocks, and thus offers the most flexibility. Meanwhile, some plugin needs to visit
+the in-memory state (like waiting Pods) that is only accessible via scheduler framework.
+
+#### Story 4
+
+A custom workload orchestrator may wish to modify the Pod prior to consideration for scheduling,
+without having to fork or alter the workload controller. The orchestrator may wish to make 
+time-varying (post-creation) decisions on Pod scheduling, perhaps to preserve scheduling constraints,
+avoid disruption, or prevent co-existence.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+- **Restricted state transition:** The `schedulingGates` field can be initialized only when a Pod is
+created (either by the client, or mutated during admission). After creation, each `schedulingGate`
+can be removed in arbitrary order, but addition of a new scheduling gate is disallowed.
+To ensure consistency, scheduled Pod must always have empty `schedulingGates`.
+This means that a client (an administrator or custom controller) cannot create a Pod that has
+`schedulingGates` populated in any way, if that Pod also specified a `spec.nodeName`. In this case,
+API Server will return a 409 conflict / validation error.
+
+    |                                              | non-nil schedulingGates⏸️ | nil schedulingGates▶️ |
+    |----------------------------------------------|-------------------------|-------------------------|
+    | unscheduled Pod<br>(nil <tt>nodeName</tt>)   | ✅ create<br>❌ update   | ✅ create<br>✅ update    |
+    | scheduled Pod<br>(non-nil <tt>nodeName</tt>) | ❌ create<br>❌ update   | ✅ create<br>✅ update    |
+
+- **New field disabled in Alpha but not scheduler extension:** In Alpha, the new Pod field is disabled
+by default. However, the scheduler's extension point is activated no matter the feature gate is enabled
+or not. This enables scheduler plugin developers to tryout the feature even in Alpha, by crafting
+different enqueue plugins and wire with custom fields or conditions.
```

Comment:
Note that the scheduler extension points are not user facing on their own. Since the beginning we considered the framework itself and all extensions we added to it later as a "refactor", which we don't typically do behind feature gates. 

I don't mind adding a separate feature gate for this extension, but I doubt it will move the needle for the reliability of the scheduler because we can't use it to gate every side refactor we will do to implement this extension.

Reply:
In Alpha, API field and the default enqueue plugin are disabled by default. So, it's just the backing extension point getting plumbed - which is basically a no-op for end-users in alpha, and only advanced scheduler developers would like to try it out. In other words, the extension point itself (the net portion enabled in alpha) is more an SDK that we basically introduce/refactor every couple of releases in scheduler framework.